### PR TITLE
[WIP][MOE] Fuse expert subgraphes together, cat weights

### PR DESCRIPTION
**Pull request by mmikolajcz**
### Details:
 - *item1*
 - *...*

### Tickets:
 - *173664*

### CODE REVIEW

**Review by mryzhov — DISMISSED** [review]
Could you please provide the description for this PR? And it would be better to add short description of this transformation into the header

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+
+                    if (decompress) {
+                        needs_decompress = true;
+                        target_type = weight->get_output_element_type(0);
+                    }
+
+                    inputs.emplace_back(ov::op::util::make_try_fold<ov::op::v0::Unsqueeze>(original_weight, const_0));
+                }
+
+                auto fused = ov::op::util::make_try_fold<ov::op::v0::Concat>(inputs, 0);
```
Comment (mryzhov) [inline src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp:398]:
```suggestion
                auto fused = std::make_shared<ov::op::v0::Concat>(inputs, 0);
                fused->get_rt_info()["postponed_constant"] = true;
```

Comment (mryzhov) [reply src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp:398]: That will solve the memory consumption issue, the constants will be merged during the serialization

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+            auto final_output = std::make_shared<ov::op::v1::ReduceSum>(weighted_outputs, axis0_vector, false);
+
+            // Reshape back to original shape and add residual connection
+            auto target_shape = std::make_shared<ov::op::v3::ShapeOf>(view_reshape_node, element::i64);
+            auto final_reshape = std::make_shared<ov::op::v1::Reshape>(final_output, target_shape, false);
+            auto final_add = std::make_shared<ov::op::v1::Add>(residual_input_node, final_reshape);
+
+            // Replace the last add node with our new computation
+            ov::replace_node(last_add_node, final_add);
+            std::cout << "MoE layer " << moe_idx + 1 << "/" << num_last_add << " fused successfully." << std::endl;
```
[the remark this answers is not in the document]
Comment (mmikolajcz) [reply src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp]: Removed debug prints

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+    if (constant->get_element_type() == element::i32) {
+        return end_value == std::numeric_limits<int32_t>::max();
+    }
+
+    if (constant->get_element_type() == element::i64) {
+        return end_value == std::numeric_limits<int64_t>::max();
+    }
+
+    return false;
+}
```
Comment (mitruska) [inline src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp:85]: The `is_slice_to_end` seems to checks the following options:
- [1, -1] - but Slice `end` is taken exclusively, so it can miss last element
- [1, MAX_I32]
- [1, MAX_I64]

Could you explain why the first value needs to be `1`?

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+    auto index_add__ScatterElementsUpdate_2 = pattern::any_input();
+    auto shape_const = wrap_type<ov::op::v0::Constant>();
+    auto expert_id = wrap_type<ov::op::v0::Constant>();
+    auto slice_end_const = wrap_type<ov::op::v0::Constant>();
+    auto select_Gather_1 = wrap_type<ov::op::v8::Gather>(
+        {permute_Transpose, expert_id, wrap_type<ov::op::v0::Constant>(pattern::value_matches("0"))},
+        {{"batch_dims", 0}});
+    auto squeeze_Squeeze_1 =
+        wrap_type<ov::op::v0::Squeeze>({select_Gather_1, wrap_type<ov::op::v0::Constant>(pattern::value_matches("0"))});
+    auto ListUnpack_NonZero_1 = wrap_type<ov::op::v3::NonZero>({squeeze_Squeeze_1}, {{"output_type", "i64"}});
```
Comment (mitruska) [inline src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp]: Is the "NonZero" `{"output_type", "i64"}` crucial here? It should work also with {"output_type", "i32"}, can it be relaxed?

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+    auto ListUnpack_Split_1 = wrap_type<ov::op::v1::Split>(
+        {ListUnpack_NonZero_1, wrap_type<ov::op::v0::Constant>(pattern::value_matches("0"))},
+        {{"num_splits", 2}});
+    ListUnpack_Split_1->set_output_size(2);
+    auto ListUnpack_Squeeze_0_1 = wrap_type<ov::op::v0::Squeeze>(
+        {ListUnpack_Split_1->output(1), wrap_type<ov::op::v0::Constant>(pattern::value_matches("0"))});
+    auto index_add__Convert_1 = wrap_type<ov::op::v0::Convert>({ListUnpack_Squeeze_0_1}, {{"destination_type", "i32"}});
```
Comment (mitruska) [inline src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp]: The `Convert` could be optional here.

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+    auto permute_Transpose = wrap_type<ov::op::v1::Transpose>({one_hot_OneHot, transpose_perm});
+    auto select_Gather = wrap_type<ov::op::v8::Gather>({permute_Transpose, axis0, axis0}, {{"batch_dims", 0}});
+    auto squeeze_Squeeze = wrap_type<ov::op::v0::Squeeze>({select_Gather, axis0});
+    auto ListUnpack_NonZero = wrap_type<ov::op::v3::NonZero>({squeeze_Squeeze}, {{"output_type", "i64"}});
+    auto ListUnpack_Split = wrap_type<ov::op::v1::Split>({ListUnpack_NonZero, axis0}, {{"num_splits", 2}});
+    ListUnpack_Split->set_output_size(2);
+    auto ListUnpack_Squeeze_0 = wrap_type<ov::op::v0::Squeeze>({ListUnpack_Split->output(1), axis0});
+    auto index_add__Convert = wrap_type<ov::op::v0::Convert>({ListUnpack_Squeeze_0}, {{"destination_type", "i32"}});
+    auto index_add__Reshape =
+        wrap_type<ov::op::v1::Reshape>({index_add__Convert, reshape_shape}, {{"special_zero", false}});
```
Comment (mitruska) [inline src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp]: Similar comment for NonZero, output type could be any and Convert optional.

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+namespace {
+
+bool is_slice_to_end(const std::shared_ptr<Node>& node) {
+    auto constant = ov::as_type_ptr<ov::op::v0::Constant>(node);
+    if (!constant || !constant->get_element_type().is_integral_number()) {
+        return false;
+    }
+
+    const auto values = constant->cast_vector<int64_t>();
+    if (values.size() != 2 || values[0] != 1) {
```
Comment (evkotov) [inline src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp:70]: could you please add some comment here why 2 and 1

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+                     down_proj_weight,
+                     permute_Transpose,
+                     slice_end_const);
+    return block;
+}
+
+}  // namespace
+
+ov::pass::FuseMOEExperts::FuseMOEExperts() : MultiMatcher("FuseMOEExperts") {
+    auto axis0 = wrap_type<ov::op::v0::Constant>(pattern::value_matches("0"));
```
Comment (evkotov) [inline src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp]: is there any way to simpilfy this code? very huge pattern and I see duplicates: axis0, ..., reshapes, ..
consider to break down into smaller parts 
// Helper methods to improve readability
  auto create_router_pattern() { /* ... */ }
  auto create_expert_indexing_pattern() { /* ... */ }
  auto create_reshape_and_add_pattern() { /* ... */ }
...

File: src/common/transformations/tests/common_optimizations/fuse_moe_test.cpp [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+    auto final_output = std::make_shared<ov::op::v1::ReduceSum>(weighted_outputs, axis0_vector, false);
+
+    auto final_reshape = std::make_shared<ov::op::v1::Reshape>(final_output, target_shape, false);
+    auto final_add = std::make_shared<ov::op::v1::Add>(residual_input, final_reshape);
+    final_add->set_friendly_name("moe_decomposed");
+
+    return std::make_shared<ov::Model>(final_add,
+                                       ov::ParameterVector{final_hidden_states_, router_logits, hidden_states_2d});
+}
+
```
Comment (evkotov) [inline src/common/transformations/tests/common_optimizations/fuse_moe_test.cpp:316]: Consider to add edge case tests:
  - Different expert counts (1, 2, large numbers like 64, 128)
  - Different topk values (1, topk==expert_num, edge cases)
  - No decompression case (fp32 weights without Convert)
  - Mixed weights (some experts with decompression, some without)
  - Invalid expert_id ordering (non-sequential IDs)
  - Missing experts (gaps in expert IDs, e.g., 0,1,3,4 - missing 2)
  - Mismatched weight dimensions between experts
  - Multiple MoE layers in the same model
  
Should we consider these test cases or they are not possible?

Comment (mmikolajcz) [reply src/common/transformations/tests/common_optimizations/fuse_moe_test.cpp:316]: Yeah, I agree that this isn't tested well enough. Also given pattern size we need to add more e2e tests to make sure that some other transformation wouldn't break matching.
As for temporary workaround I've added e2e test using tiny models here: tests/model_hub_tests/transformation_tests/test_moe_transformation.py
It is not an ideal solution but it looks for specific pattern of Matmul, checks for optional converts and executes some basic accuracy tests.

As for proper testing I fear this would take more time to develop so maybe it could be considered as follow up

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp (hunk 7 above)
Comment (mmikolajcz) [reply src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp:70]: It was used to catch specific slice end found in model like from previous MoE decomposition. I've added additional docstring in code
 https://github.com/openvinotoolkit/openvino/pull/30448/files#diff-84784725bb2c5e2c496ed577c2d5f841074540a7c619aef527e5006ba388159fR99

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp (hunk 3 above)
Comment (mmikolajcz) [reply src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp:85]: It was used to catch specific slice end found in model like from previous MoE decomposition where slice end was set to specifically [1, MAX]. I've added additional docstring in code
 https://github.com/openvinotoolkit/openvino/pull/30448/files#diff-84784725bb2c5e2c496ed577c2d5f841074540a7c619aef527e5006ba388159fR99

+ removed -1 case since it was incorrect and for now just MAX val is needed

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp (hunk 8 above)
Comment (mmikolajcz) [reply src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp]: Simplified this code by creating functions for characteristic parts of MOE. In future, maybe this could also help with creating more generalized pattern to handle different models

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp (hunk 6 above)
Comment (mmikolajcz) [reply src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp]: Done

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp (hunk 5 above)
Comment (mmikolajcz) [reply src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp]: Done

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp (hunk 4 above)
Comment (mmikolajcz) [reply src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp]: Done

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp [hunk 10]
```diff
@@ -0,0 +1,535 @@
...
+    }
+
+    if (constant->get_element_type() == element::i64) {
+        return end_value == std::numeric_limits<int64_t>::max();
+    }
+
+    return false;
+}
+
+struct expert_data {
```
Comment (mryzhov) [inline src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp]: I would prefer to define this structure in the header, then you can use it in the tests

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp [hunk 11]
```diff
@@ -0,0 +1,535 @@
...
+    std::shared_ptr<Node> gate_proj_weight;
+    std::shared_ptr<Node> up_proj_weight;
+    std::shared_ptr<Node> down_proj_weight;
+    size_t expert_id;
+    std::shared_ptr<Node> permute_node;
+};
+
+// Pattern matching for individual expert computation block with SwiGLU activation
+std::shared_ptr<pattern::op::Block> mlp3_no_bias_swiglu_block(
+    const Output<Node>& permute_Transpose,    // Transpose -> OneHot -> TopK -> Softmax -> MatMul -> Hidden States
```
Comment (mryzhov) [inline src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp:89]: Do we really need to detect transpose, gather, reshape, squeeze and unsqueeze layers in the pattern?

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp [hunk 12]
```diff
@@ -0,0 +1,535 @@
...
+                continue;
+            }
+
+            auto& experts = experts_it->second;
+            std::sort(experts.begin(), experts.end(), [](const expert_data& a, const expert_data& b) {
+                return a.expert_id < b.expert_id;
+            });
+
+            // Build fused weight tensors by concatenating individual expert weights along expert dimension
+            auto const_0 = ov::op::v0::Constant::create(element::i64, Shape{1}, {0});
```
Comment (mryzhov) [inline src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp]: Looks like we can create this Constant once for all MOE layers

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp [hunk 13]
```diff
@@ -0,0 +1,533 @@
...
+    auto axis_scalar = ov::op::v0::Constant::create(element::i64, Shape{}, {static_cast<int64_t>(axis_index)});
+    auto axis0_scalar = ov::op::v0::Constant::create(element::i64, Shape{}, {0});
+    auto dim_scalar = std::make_shared<ov::op::v8::Gather>(shape_of, axis_scalar, axis0_scalar);
+    auto axis0_vector = ov::op::v0::Constant::create(element::i64, Shape{1}, {0});
+    return std::make_shared<ov::op::v0::Unsqueeze>(dim_scalar, axis0_vector);
+}
+
+// Helper function to create router pattern (Softmax -> TopK -> OneHot -> Transpose)
+// This pattern identifies the expert selection logic from router logits
+std::pair<std::shared_ptr<Node>, std::shared_ptr<Node>> create_router_pattern(const AxisPatterns& axes) {
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp]: The input parameter `axes` is not used. Do we need it here?

Comment (mmikolajcz) [reply src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp]: Thanks, removed

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp (hunk 10 above)
Comment (mmikolajcz) [reply src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp]: Moved to header

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp (hunk 11 above)
Comment (mmikolajcz) [reply src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp:89]: I think this could be refactored to remove this subpattern, however at current state replacing this op chain with pattern::any_input causes callback to throw, probably due to removing connection to hidden_states shape used in callback reshapes.
This information could be retrieved from different part of pattern, however I'm afraid this change might require more effort so to not block it this could be done as follow-up in 2026.0 release

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp
Comment (chenhu-wang) [inline src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp:1]: @mmikolajcz  In below converted qwen3 IR, we see the Add operation has incompatible input shapes. One is [?, 2048], another is [?,?,2048]. Could you please help check?
image

image

Comment (mmikolajcz) [reply src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp:1]: Just to be sure, second screen is after pass to fuse MOE subgraph to internal MOE? As for transformation to weight fusion I've tested inference and there were no mismatches in shape

File: src/common/transformations/tests/common_optimizations/fuse_moe_test.cpp [hunk 14]
```diff
@@ -0,0 +1,439 @@
...
+    auto axis_minus_one = makeConst(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{-1});
+    auto sum_ReduceSum = makeOP<opset1::ReduceSum>({topk_TopK->output(0), axis_minus_one}, {{"keep_dims", true}});
+    auto div__Divide = makeOP<opset1::Divide>({topk_TopK->output(0), sum_ReduceSum},
+                                              {{"auto_broadcast", "numpy"}, {"m_pythondiv", true}});
+    auto one_hot_OneHot = makeOP<opset1::OneHot>({topk_TopK->output(1), expert_num, 1, 0}, {{"axis", 2}});
+    // param2: expert_mask: [128, 8, batch]
+    auto permute_Transpose = makeOP<opset1::Transpose>({one_hot_OneHot, {2, 1, 0}});
+
+    // hidden_states_2d: f32[-1, 2048]
+    // auto view_Reshape = makeOP(ov::Rank(2));
```
Comment (CuriousPanCake) [inline src/common/transformations/tests/common_optimizations/fuse_moe_test.cpp:46]:
```suggestion
```

File: src/common/transformations/tests/common_optimizations/fuse_moe_test.cpp [hunk 15]
```diff
@@ -0,0 +1,439 @@
...
+    for (int i = 0; i < expert_num; i++) {
+        // expert_mask[expert_idx]
+        std::shared_ptr<Node> select_Gather_2;
+
+        select_Gather_2 = makeOP<opset8::Gather>({expert_mask, i, 0}, {{"batch_dims", 0}});
+        auto squeeze_Squeeze_7 = makeOP<opset1::Squeeze>(
+            {select_Gather_2,
+             squeeze_axis_const});  //  tensor_array<i64[2,?]>
+                                    //  __module.model.layers.1.mlp/aten::squeeze/Squeeze_7(__module.model.layers.1.mlp/aten::select/Gather_7,
+                                    //  60)
```
Comment (CuriousPanCake) [inline src/common/transformations/tests/common_optimizations/fuse_moe_test.cpp:92]: Clean-up to do

File: src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp
Comment (chenhu-wang) [reply src/common/transformations/src/transformations/common_optimizations/fuse_moe_experts.cpp:1]: @mmikolajcz, yes, second pic is to fused all MOE surroundings(softmax,topk...), to more clear show the shape propagation. "Add" operation has input shapes [?, 2048] and [?,1,2048] in converted IR, while "?" is seq_len, could be value other than 1, and can not be broadcasted.
